### DISCUSSION TITLE
Why not render again?

### DISCUSSION THREAD

**Discussion by Pulset** [pmndrs/valtio#824] [Q&A]
I changed the state, but I didn't render it again.Is this a bug or is this my misunderstanding?

This is my demo:
https://codesandbox.io/s/crazy-lewin-7rpy8w?file=/src/useListConfig.js

**Comment by dai-shi** [comment]
If I don't misunderstand, you simply assign the same object. `useSnapshot` is smart enough so that it optimizes re-renders. `name` and `property.value` don't change.

**Reply by Pulset** [reply]
The proxy tracks changes to the original object and all nested objects, notifying listeners when an object is modified.If I assign property a new object,it will not re-renders.
```js
menuItems.forEach((item) => {
    if (item.menuItems) {
      item.menuItems.forEach((menuItem) => {
        if (menuItem.property && menuItem.property.key) {
          // assign a new object
          menuItem.property = {
            vlaue: statusValuesMap[menuItem.property.key],
            key: menuItem.property.key
          };
        }
      });
    }
  });
```

**Reply by Pulset** [reply]
When the entire object and nested objects return new objects,it will re-render.Does this mean that only changing nested objects cannot re-render?

**Reply by Aslemammad** [reply]
Hey Pulset! 

Valtio is so smart at optimizing renders.
if you're tracking the `property.key` (as a string) and you change `property`'s reference, or assign a new value to it, it would not cause re-render, since valtio knows that you care about only `property.key`, and not `property` itself or `property.value` as a  whole.

Here's an explanation for your comments, I hope it helps you.

```ts
    // why not re-render ?
    // state[key] = [] will re-render: This will change menuItems (removes it), since valtio knows you care about this, it will re-render
    // state[key] = initData will not re-render: This is not changing anything, menuItems, menu.name and menu.property.value stay the same
    // state[key] = [...initData] will not re-render: same, valtio knows you care about menuItems for instance, but you're not changing menuItems and you're only changing the parent object reference. 
    // state[key] = JSON.parse(JSON.stringify(initData)): This completely destroys the references recursively. but if the tracked values are primitives only, nothing would happen.

```

**Reply by Aslemammad** [reply]
Let me know if this helps, and mark it as an answer if it did so other people would benefit from it.

If you have any other question, feel free to ask here.

**Reply by Aslemammad** [reply]
I just had a mistake in the comments that I just updated right now. Thank you @dai-shi for pointing this out.

**Comment by Pulset** [comment]
> Hey Pulset!
> 
> Valtio is so smart at optimizing renders. if you're tracking the `property.key` (as a string) and you change `property`'s reference, or assign a new value to it, it would not cause re-render, since valtio knows that you care about only `property.key`, and not `property` itself or `property.value` as a whole.
> 
> Here's an explanation for your comments, I hope it helps you.
> 
> ```ts
>     // why not re-render ?
>     // state[key] = [] will re-render: This will change menuItems (removes it), since valtio knows you care about this, it will re-render
>     // state[key] = initData will not re-render: This is not changing anything, menuItems, menu.name and menu.property.value stay the same
>     // state[key] = [...initData] will not re-render: same, valtio knows you care about menuItems for instance, but you're not changing menuItems and you're only changing the parent object reference. 
>     // state[key] = JSON.parse(JSON.stringify(initData)) will re-render: This completely destroys the references recursively.
> ```
// state[key] = initData will not re-render: This is not changing anything, menuItems, menu.name and menu.property.value stay the same
The original object has no value attribute and I care about this.Then I add the value attribute by assigning a brand new object.So it's different from the original.

**Reply by Aslemammad** [reply]
![image](https://github.com/pmndrs/valtio/assets/37929992/ea5cc14e-3a6c-4a35-a8e9-505d8be8d575)
https://codesandbox.io/s/crazy-lewin-7rpy8w?file=/src/useListConfig.js:919-961

I don't see it empty before the assignment.

And please let me know where you think it was empty. And when do you want it to re-render?

**Reply by Pulset** [reply]
<img width="561" alt="image" src="https://github.com/pmndrs/valtio/assets/21022593/b14e8ae7-7f54-4ee8-b4ac-1f2231e961b5">


this property object is different

**Reply by Aslemammad** [reply]
```tsx
 <p>{menu.property.value}</p>
```

You're tracking `value` and not `vlaue`!

**Reply by Pulset** [reply]
It's my mistake, I reworked it.But it still can't re-render

<img width="509" alt="image" src="https://github.com/pmndrs/valtio/assets/21022593/fd8abcd9-9d78-4ac8-9b87-ed89cb3f5ece">

**Reply by Aslemammad** [reply]
@Pulset I just discussed this with Daishi, and he told me the reason. 

> This is because valtio caches proxies based initial objects. This is intentional design to leave initial objects unmodified (= proxy() is a pure function). Thus, modifying initial objects isn't recommended. This has been reported several times, and while the current design helps misusing, we can't opt-out the behavior. So, in v2, we changed the design and we don't pass initial objects, but we pass base objects instead, that will be modified (= proxy() is an impure function). 
https://codesandbox.io/s/still-flower-q7jxj9?file=/src/useListConfig.js:1304-1361 Here's the example with v2.0.0-alpha.0.
